kin-devplatform-sdk/src/main/java/kin/devplatform/KinEcosystemInitiator.java

```diff
@@ -231,33 +217,48 @@ public void onError(Exception e) {
 		});
 	}
 
-	private void handleKinClientReady(IKinClient kinClient, Context context, String appId, SignInData signInData,
+	private void handleKinClientReady(IKinClient kinClient, MigrationManager migrationManager,
+		Context context, String appId, SignInData signInData,
 		boolean withLogin, KinCallback<Void> loginCallback) throws BlockchainException {
 		final EventLogger eventLogger = EventLoggerImpl.getInstance();
-		BlockchainSourceImpl.init(eventLogger, kinClient, BlockchainSourceLocal.getInstance(context));
-		if (appId != null) {
-			BlockchainSourceImpl.getInstance().setAppID(appId);
-		}
-
-		AccountManagerImpl
-			.init(AccountManagerLocal.getInstance(context), eventLogger, AuthRepository.getInstance(),
-				BlockchainSourceImpl.getInstance());
-
-		OrderRepository.init(BlockchainSourceImpl.getInstance(),
-			eventLogger,
-			OrderRemoteData.getInstance(executorsUtil),
-			OrderLocalData.getInstance(context, executorsUtil));
-
-		OfferRepository
-			.init(OfferRemoteData.getInstance(executorsUtil), OrderRepository.getInstance());
+		// If not initialized then initialize all necessary components.
+		if (!isInitialized) {
+			BlockchainSourceImpl.init(eventLogger, kinClient, BlockchainSourceLocal.getInstance(context));
+			if (appId != null) {
+				BlockchainSourceImpl.getInstance().setAppID(appId);
+			}
 
-		DeviceUtils.init(context);
+			AccountManagerImpl
+				.init(AccountManagerLocal.getInstance(context), migrationManager, eventLogger,
+					AuthRepository.getInstance(),
+					BlockchainSourceImpl.getInstance());
+
+			OrderRepository.init(BlockchainSourceImpl.getInstance(),
+				eventLogger,
+				OrderRemoteData.getInstance(executorsUtil),
+				OrderLocalData.getInstance(context, executorsUtil));
+
+			OfferRepository
+				.init(OfferRemoteData.getInstance(executorsUtil), OrderRepository.getInstance());
+
+			DeviceUtils.init(context);
+			isInitialized = true;
+			if (withLogin) {
+				eventLogger.send(KinSdkInitiated.create());
+				login(signInData, loginCallback);
+				return;
+			}
+		} else {
+			// if already initialized then we only did the migration so only update the kinClient and account.
+			BlockchainSourceImpl.getInstance().updateKinClientAndAccount(kinClient, kinClient.getAccountCount() - 1);
```

Comment:
should also re-register listeners (like in `updateActiveAccount`) cause now you can be in kin3

Reply:
Fixed